Alpha JWS Discovery API for locating an apiserver securely

dgoodwin opened the pull request:
This PR contains an early alpha prototype of the JWS discovery API outlined in proposal #30707.

CA certificate, API endpoints, and the token to be used to authenticate to this discovery API are currently passed in as secrets. If the caller provides a valid token ID, a JWS signed blob of ClusterInfo containing the API endpoints and the CA cert to use will be returned to the caller. This is used by the alpha kubeadm to allow seamless, very quick cluster setup with simple commands well suited for copy paste.

Current TODO list:
- [x] Allow the use of arbitrary strings as token ID/token, we're currently treating them as raw keys.
- [x] Integrate the building of the pod container, move to cluster/images/kube-discovery.
  - [x] Build for: amd64, arm, arm64 and ppc64le. (just replace GOARCH=)
  - [x] Rename to gcr.io/google_containers/kube-discovery-ARCH:1.0
  - [x] Cleanup rogue files in discovery sub-dir.
  - [x] Move pkg/discovery/ to cmd/discovery/app.

There is additional pending work to return a kubeconfig rather than ClusterInfo, however I believe this is slated for post-alpha.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32203)

<!-- Reviewable:end -->

dgoodwin committed:
Add standalone kube-discovery JWS discovery API.

This is a standalone pod which will be configured by kubeadm for the
time being. A token ID/token map, endpoints list, and CA cert are
provided as secrets.

Callers request the cluster info by shared secret (token ID), and if the
token ID matches a JWS signed payload is returned using the other half
of the shared secret to validate.

errordeveloper committed:
Refactor and provide JSON file loaders for endpoints and tokens

dgoodwin committed:
Expand on kube-discovery API and integrate container build.

luxas commented:
@kubernetes/sig-cluster-lifecycle

luxas commented on discovery/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM golang
```
This shouldn't be in the `discovery/` dir, but `cluster/images/jws-discovery` or something like that.

Also, it should be 
```
FROM busybox
COPY kube-discovery /usr/local/bin
ENTRYPOINT ["/usr/local/bin/kube-discovery"]
````

in order to be consistent with the other server images

luxas commented on discovery/kubediscovery.yaml:
```diff
@@ -0,0 +1,37 @@
+      hostNetwork: true
+      containers:
+      - name: kubediscovery
+        image: dgoodwin/kubediscovery
```
update repo

dgoodwin commented:
Introduced a Makefile, is this on track for building the container for appropriate arches @luxas ?

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,41 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Build the hyperkube image.
```
kube-discovery

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,41 @@
+REGISTRY?=gcr.io/google_containers
+ARCH?=amd64
+TEMP_DIR:=$(shell mktemp -d)
+
```
Replace BASEIMAGE here based on ARCH like in the hyperkube image
amd64: busybox
arm: armel/busybox
arm64: aarch64/busybox
ppc64le/busybox

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,41 @@
+all: build
+
+build:
+ifndef VERSION
```
Please set version to 1.0 initially

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,41 @@
+endif
+	cp -r ./* ${TEMP_DIR}
+	cp ../../../_output/dockerized/bin/linux/${ARCH}/kube-discovery ${TEMP_DIR}
+	sudo docker build -t ${REGISTRY}/kube-discovery-${ARCH}:${VERSION} ${TEMP_DIR}
```
Remove sudo

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,41 @@
+    $(error VERSION is undefined)
+endif
+	cp -r ./* ${TEMP_DIR}
+	cp ../../../_output/dockerized/bin/linux/${ARCH}/kube-discovery ${TEMP_DIR}
```
document that to create this one should run `build/run.sh /bin/bash -c "KUBE_BUILD_PLATFORMS=linux/ARCH make WHAT=cmd/kube-discovery"`

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,41 @@
+
+push: build
+	gcloud docker push ${REGISTRY}/kube-discovery-${ARCH}:${VERSION}
+	docker tag -f ${REGISTRY}/kube-discovery-${ARCH}:${VERSION} ${REGISTRY}/kube-discovery:${VERSION}
```
Remove the old naming, we don't want to have use that

dgoodwin replied:
Sorry what is the old/new naming you refer to?

dgoodwin replied:
Via slack, just removing the tagging and push of a "kube-discovery" image.

luxas commented on cluster/images/kube-discovery/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM busybox
```
FROM BASEIMAGE

luxas commented on cluster/images/kube-discovery/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+COPY kube-discovery /usr/local/bin
+ENTRYPOINT "/usr/local/bin/kube-discovery"
+
+EXPOSE 8080
```
Not necessary, but not against it

dgoodwin replied:
Hmm I think it's actually grabbing 9898, so this is dead wrong, I will remove.

luxas commented:
Please nuke the Makefile and cluster/images/hyperkube/Makefile changes...

dgoodwin commented:
Hyperkube makefile changes were a mis-commit, but the Makefile cross change was part of enabling building specific binaries instead of the world with make cross, does this one seem ok to bring to main repo?

luxas commented:
The makefile change isn't required, see: `build/run.sh /bin/bash -c "KUBE_BUILD_PLATFORMS=linux/ARCH make WHAT=cmd/kube-discovery"`

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,55 @@
+endif
+ifeq ($(ARCH),arm)
+	BASEIMAGE?=armel/debian:jessie
+	QEMUARCH=arm
```
Qemuarch not needed in this case

luxas commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,55 @@
+all: build
+
+build:
+	VERSION?=1.0
```
This should be moved up with the other variables

luxas commented on cluster/images/kube-discovery/README.md:
````diff
@@ -0,0 +1,39 @@
+## Build And Run From Source
+
+```
+$ make WHAT=cmd/kube-discovery
````
Update this to the command I gave earlier

dgoodwin replied:
Updated, I swapped in amd64 as that seems overwhelmingly most likely for a person looking to build from source and run it locally.

luxas replied:
Great!

mikedanese commented:
We will have both:
```
cmd/kubernetes-discovery
cmd/kube-discovery
```
after this....

dgoodwin commented:
@luxas proposed jws-discovery, sound ok to everyone?

luxas commented on cluster/images/kube-discovery/README.md:
```diff
@@ -0,0 +1,39 @@
+
+This image is published temporarily on Docker Hub as dgoodwin/kubediscovery
+
+`docker run --rm -p 9898:9898 -v /tmp/secret/ca.pem:/tmp/secret/ca.pem --name kubediscovery dgoodwin/kubediscovery`
```
This should be replaced with the "real" name

luxas replied:
Why not append `-d` as well?

luxas commented on cluster/images/kube-discovery/README.md:
```diff
@@ -0,0 +1,39 @@
+
+## Testing the API
+
+`curl "http://localhost:9898/cluster-info/v1/?token-id=TOKENID"`
```
Can you define what a valid token is and where it comes from. To me, it's not obvious here.
But I do not want to block on docs like this initially...

dgoodwin replied:
The token ID and token are just random strings, we're not doing much validation on them today. Both are generated in kubeadm and then passed through to the pod as a secret.

luxas commented on cluster/images/kube-discovery/README.md:
```diff
@@ -0,0 +1,39 @@
+
+`curl "http://localhost:9898/cluster-info/v1/?token-id=TOKENID"`
+
+You should see JSON containing a signed payload. For code to verify and decode that payload see handler_test.go.
```
I think you should run `update-generated-docs` and `update-munge-docs`

luxas reviewed (commented):
Did a separate review round.
Still, if we settle on `jws-discovery`, everything needs to be renamed

luxas commented:
Oh, this was SO much easier to review now when go-jose is gone. +1

luxas commented:
@dgoodwin If you also can squash your commits to 1-3 it would be great!

dgoodwin commented:
Updated README to better clarify how the secrets are passed through, and how to run as a Docker container using the container locations we expect it will be published at. (gcr.io/google_containers/kube-discovery)

Ran the update docs scripts and committed the result. 

Squashed down to 3 commits so I can preserve @errordeveloper 's contribution.

luxas commented:
Still the boilerplates are wrong:
```
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/cmd/kube-discovery/kubediscovery.go
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/cmd/kube-discovery/app/model.go
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/cmd/kube-discovery/app/handlers_test.go
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/cmd/kube-discovery/app/routes.go
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/cmd/kube-discovery/app/handlers.go
Boilerplate header is wrong for: /go/src/k8s.io/kubernetes/cluster/images/kube-discovery/Dockerfile
FAILED   hack/make-rules/../../hack/verify-boilerplate.sh	1s
```

luxas commented:
Lastly, we should apply `cla: human-approved` I think.
I guess @thockin has the legal power to do that :)

dgoodwin commented:
Boilerplate updated.

Hmm, I should be CLA approved, at least I was for this: https://github.com/kubernetes/release/pull/50, via being added to the kubernetes-redhat-contributors group.

luxas commented:
Yes, I just think the issue is that the CLA bot doesn't allow two committers

luxas commented:
Ping @jbeda @errordeveloper @mikedanese @smarterclayton 
This PR is ready for final review.

@mikedanese Should build and push the images as soon as it's LGTM'd (which I hope is **today**)

taimir commented:
CC @cheld @kenan435 @floreks

taimir commented on cluster/images/kube-discovery/Makefile:
```diff
@@ -0,0 +1,52 @@
+all: build
+
+build:
+	cp -r ./* ${TEMP_DIR}
```
@dgoodwin I think you need a line similar to this here:
```{bash}
cd ${TEMP_DIR} && sed -i.back "s|BASEIMAGE|${BASEIMAGE}|g" Dockerfile
```

At least that's how it's done in the `hyperkube` image :)

dgoodwin replied:
Added, thanks!

errordeveloper commented:
LGTM, would be great to get this in as soon as we can!

lukemarsden commented:
Friendly ping @mikedanese :)

dgoodwin commented:
@mikedanese I have no preference on how that image gets built so say the word and I can apply https://storage.googleapis.com/mikedanese/patches/0001-build-kube-discovery-image-with-release.patch if you like. It does look simpler.

luxas commented:
I still think we should follow the kube-dns approach with tagged `1.x` releases that is built/pushed whenever it wants to, instead of following the 2 week alpha schedule (which in the worst case, v1.4.0-alpha.3, got two more weeks delayed)

Everything for that is done. @mikedanese I vote for merging this now.

We can take this discussion again when we start thinking about promoting it to beta.

luxas commented:
@k8s-bot test this please github issue: #IGNORE

luxas commented:
@mikedanese Please push these images for all architectures today!
I think that would make many sig-cluster-lifecycle guys happy :)

errordeveloper commented:
Amazing to see this merged! Please ping me whenever we have images on GCR, and I'll update #33262 accordingly.

lukemarsden commented:
@mikedanese do we have images on GCR for this?

luxas commented:
Yes
